File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TankDrive.java
package org.firstinspires.ftc.teamcode;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.DcMotor;

@TeleOp(name = "Tank Drive")
public class TankDrive extends LinearOpMode {
    private Hardware hw;

    private void runDrivetrain() {
        hw.backLeftDrive.setPower(-gamepad1.left_stick_y);
        hw.frontLeftDrive.setPower(-gamepad1.left_stick_y);
        hw.backRightDrive.setPower(-gamepad1.right_stick_y);
        hw.frontRightDrive.setPower(-gamepad1.right_stick_y);
    }

    private void runArm() {
        double armPower = -gamepad2.right_stick_y;
        double extensionPower = -gamepad2.left_stick_y;

//        int newPosition = (int) (hw.leftArm.getCurrentPosition() + (50 * armPower));
//        if (Math.abs(armPower) > 0.1) {
//            hw.setArmPosition(newPosition);
//        }
//        hw.leftArm.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
//        hw.rightArm.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
//        hw.leftArm.setPower(armPower);
//        hw.rightArm.setPower(armPower);
        telemetry.addData("Arm position", hw.leftArm.getCurrentPosition());
        telemetry.addData("Arm position right", hw.rightArm.getCurrentPosition());

//        hw.armExtender.setPower(extensionPower);
//        telemetry.addData("Arm extender position", hw.armExtender.getCurrentPosition());
    }

    private void runGrabber() {
        double grabberPos = gamepad2.right_trigger;
        double grabberPivotPos = gamepad2.left_trigger;

        hw.grabberPivot.setPosition(grabberPivotPos);
        hw.grabber.setPosition(grabberPos);
        telemetry.addData("Grabber position", hw.grabber.getPosition());
        telemetry.addData("Grabber pivot pos", hw.grabberPivot.getPosition());

//        hw.armHolder.setPosition(gamepad2.a ? Hardware.ARM_HOLDER_RETRACTED : Hardware.ARM_HOLDER_HOLDING);
    }



    @Override
    public void runOpMode() throws InterruptedException {
        hw = new Hardware(hardwareMap);

        telemetry.addData("Status", "Initialized");
        telemetry.update();

        waitForStart();

        while (opModeIsActive()) {
            runDrivetrain();
            runArm();
            runGrabber();

            hw.led.setPower(gamepad2.x ? 0.5 : 0);

            telemetry.update();
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/TankDrive.kt
package org.firstinspires.ftc.teamcode

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.util.ElapsedTime
import kotlin.math.roundToInt

/**
 * This file contains an minimal example of a Linear "OpMode". An OpMode is a 'program' that runs in either
 * the autonomous or the teleop period of an FTC match. The names of OpModes appear on the menu
 * of the FTC Driver Station. When an selection is made from the menu, the corresponding OpMode
 * class is instantiated on the Robot Controller and executed.
 *
 * This particular OpMode just executes a basic Tank Drive Teleop for a two wheeled robot
 * It includes all the skeletal structure that all linear OpModes contain.
 *
 * Use Android Studios to Copy this Class, and Paste it into your team's code folder with a new name.
 * Remove or comment out the @Disabled line to add this opmode to the Driver Station OpMode list
 */

@TeleOp(name = "Tank Drive", group = "Linear Opmode")
class TankDrive : LinearOpMode() {
    private val runtime = ElapsedTime()
    private lateinit var hw: Hardware

    override fun runOpMode() {
        hw = Hardware(hardwareMap, this)

        telemetry.addData("Status", "Initialized")
        telemetry.update()

        waitForStart()
        runtime.reset()

        // run until the end of the match (driver presses STOP)
        val loopTime = ElapsedTime()

        while (opModeIsActive()) {
            loopTime.reset()

            hw.backLeftDrive.power = gamepad1.left_stick_y.toDouble()
            hw.frontLeftDrive.power = gamepad1.left_stick_y.toDouble()
            hw.backRightDrive.power = gamepad1.right_stick_y.toDouble()
            hw.frontRightDrive.power = gamepad1.right_stick_y.toDouble()

            // Show the elapsed game time
            telemetry.addData("Status", "Run Time: $runtime")
            telemetry.addData("Loop time", loopTime.milliseconds().toString() + "ms")
            telemetry.update()
        }
    }
}
